### PR TITLE
fix: does not remove nulls in plugin configs

### LINKED ISSUE #3908: Cannot use nulls in fields of kongplugin.config in dbless mode
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When KIC is running in dbless mode, fields in `kongPlugin.Config` with `null` value will be omitted when sending configuration to kong: 
https://github.com/Kong/kubernetes-ingress-controller/blob/1a45c0c10ef1f4c26fe63a8e3a0ab866fe7a6c01/internal/dataplane/sendconfig/inmemory.go#L52-53
The code was first introduced in PR https://github.com/Kong/kubernetes-ingress-controller/pull/484. 
If KIC is running in db mode, null values could be present in Kong. 
This will make users unable to explicitly configure fields of plugins to `null`, which is meaningful in some plugins. For example, `datadog` plugin would use `null` in host, to load host from env variables. If KIC is running in dbless mode, and the `config.host` is set to `null`, the field is omitted and Kong will treat `host` as not specified and fill in a default value `localhost`. 

### Expected Behavior

When I explicitly specify some fields of `KongPlugin.Config` to `null`, the `null` value of the field should be passed to kong.

### Steps To Reproduce

```markdown
1. create a KongPlugin: 
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: datadog
  namespace: default
  annotations:
    kubernetes.io/ingress.class: kong
plugin: datadog
config:
  host: null
  port: 8125
2. Create an ingress with annotation: konghq.com/plugins: "datadog"
3. call /plugins admin API, and the `config.host` becomes `localhost`.
```

### Kong Ingress Controller version

```shell
v2.9.3
```

### Kubernetes version

```shell
v1.26.3
```

### Anything else?

https://github.com/Kong/kubernetes-ingress-controller/pull/484/files firstly introduces the cleanup of `null` fields in Dec 2019 and first released in KIC v0.7.0. The comment tells us Kong will return error when fields with `null` are present in config. Please help me to know more about the context at that time.

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

Do not remove `nil` values in `config` of plugins, to allow users to explicitly set some fields to `null` in plugin configs, for example `host` in `datadog` plugin.

**Which issue this PR fixes**:

fixed #3908 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Review by pmalek — COMMENTED** [review]
👍 Can you add a test for this? Ideally a simple UT where e.g. we provide our own mock of `ConfigService`

**Comment by randmonkey** [comment]
> +1 Can you add a test for this? Ideally a simple UT where e.g. we provide our own mock of `ConfigService`

Added. Also I did a simple test in my local dev env, we successfully configured `host` to `null` in `datadog` plugin.

File: internal/dataplane/sendconfig/inmemory_test.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+				Plugins: []file.FPlugin{tc.plugin},
+			}
+			_, _, err := s.Update(ctx, targetState)
+			require.NoError(t, err)
+
+			currentState, err := m.getCurrentState()
+			require.NoError(t, err)
+			plugin, err := extractPluginInFileContentByID(currentState, *tc.plugin.ID)
+			require.NoError(t, err)
+			require.True(t, reflect.DeepEqual(plugin.Config, tc.plugin.Config))
```
Comment (pmalek) [inline internal/dataplane/sendconfig/inmemory_test.go:109]: require.Equal wouldn't work here? If not then in any way we're not printing the diff incase of a failure

Comment (randmonkey) [reply internal/dataplane/sendconfig/inmemory_test.go:109]: `require.Equal` would work for strings and nils here. Updated.

**Review by pmalek — APPROVED** [review]

**Review by ms2008 — COMMENTED** [review]
The risks of this fix outweigh the benefits, because Kong's plugin schema is not guaranteed to be very robust yet, and there is a high probability that a 500 error will be triggered if a null is passed through. For example, this one: https://github.com/Kong/kong-ee/pull/4882

From the user's perspective, I think this behavior is a breaking change.

**Comment by pmalek** [comment]
> The risks of this fix outweigh the benefits, because Kong's plugin schema is not guaranteed to be very robust yet, and there is a high probability that a 500 error will be triggered if a null is passed through. For example, this one: https://github.com/Kong/kong-ee/pull/4882
> 
> From the user's perspective, I think this behavior is a breaking change.

Is there a list of known plugins, keys that could accept nils so that we could do a filtered nil clean up of config?

**Comment by randmonkey** [comment]
> The risks of this fix outweigh the benefits, because Kong's plugin schema is not guaranteed to be very robust yet, and there is a high probability that a 500 error will be triggered if a null is passed through. For example, this one: [Kong/kong-ee#4882](https://github.com/Kong/kong-ee/pull/4882)
> 
> From the user's perspective, I think this behavior is a breaking change.

@ms2008 @pmalek  For whether allowing nulls in configs of kong plugins, KIC has different behaviors: `nil`s are cleared when using dbless to directly marshal configurations, but passed to kong when using db and configuring through decK. I did not see any special processing on `null` or `nil` values in decK. 
Could we confirm again the two things:
- How did decK process config of plugins
- What process should KIC take for `null`/`nil`s in plugin config:
  (1) clean up all `nil`s
  (2) pass all `nil`s and marshal them to `null`s
  (3) Allow only explicitly specified fields, clean up `nil`s for other fields
  (4) Clean up only  explicitly specified fields, pass `nil`s for other fields

**Comment by ms2008** [comment]
@pmalek I don't think there is such a list yet. It's probably not just plugins, but other entities will have this problem too, I guess.

And seems it's better that the KIC team needs to sync up with the gateway team about this null thing.

**Comment by vm-001** [comment]
Is there any place to warn this as it may be a potential breaking change for users? Let's say there is an old KIC config and contains `some-key: nil`, it may be rejected by the Kong Admin API once KIC upgraded.

**Comment by ms2008** [comment]
There is another issue. KIC sends requests to Kong occurs in two places:

1. Admission webhook
2. Sendconfig

In the admission webhook, KIC does not remove any null values, whereas in the sendconfig, it does. These different behaviors can easily cause confusion for users. 

If you are sure you want to remove the null values, then remove them from both places; otherwise, keep them in both places. We should always ensure that the behavior of these two places is consistent.

**Comment by pmalek** [comment]
Added a "Work in Progress" label since that will require more effort to get a proper fix.

**Comment by pmalek** [comment]
Closing this as #3914 has been closed as "Not Planned".

cc: @randmonkey
